modules/src/ics07_tendermint/client_def.rs

```
@@ -25,17 +36,117 @@ impl ClientDef for TendermintClient {
 
     fn check_header_and_update_state(
         &self,
+        ctx: &dyn ClientReader,
+        client_id: ClientId,
         client_state: Self::ClientState,
         header: Self::Header,
-    ) -> Result<(Self::ClientState, Self::ConsensusState), Error> {
-        if client_state.latest_height() >= header.height() {
-            return Err(Error::low_header_height(
-                header.height(),
-                client_state.latest_height(),
-            ));
-        }
+    ) -> Result<(Self::ClientState, Self::ConsensusState), Ics02Error> {
+        // check if a consensus state is already installed; if so it should
+        // match the untrusted header.
+
+        if let Some(cs) = ctx.consensus_state(&client_id, header.height()) {
+            //could the header height be zero ?
+            let consensus_state = downcast!(
+                cs => AnyConsensusState::Tendermint
+            )
+            .ok_or_else(|| Ics02Error::client_args_type_mismatch(ClientType::Tendermint))?;
+
+            if consensus_state != ConsensusState::from(header.clone()) {
+                //freeze the client and return the installed consensus state
+                return Ok((
+                    client_state.with_set_frozen(header.height()),
+                    consensus_state,
+                ));
```

We should freeze the client only if the conflicting header verifies. i.e. we need to run the checks/ validity predicates (from the new tendermint-rs verify function) below. If the header in update doesn't verify we should return an error.